src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,678 @@
+        serviceIterator = null;
+    }
+
+    /**
+     * Instantiates a {@code KDF} object. This constructor is called when a
+     * provider is not supplied to {@code getInstance}.
+     *
+     * @param firstPairOfSpiAndProv the delegate
+     * @param t the service iterator
+     * @param algorithm the algorithm
+     * @param kdfParameters the algorithm parameters
+     */
+    private KDF(Delegate firstPairOfSpiAndProv, Iterator<Service> t, String algorithm,
+                KDFParameters kdfParameters) {
+        this.firstPairOfSpiAndProv = firstPairOfSpiAndProv;
+        serviceIterator = t;
+        this.algorithm = algorithm;
+        this.kdfParameters = kdfParameters;
+    }
+
+    /**
+     * Returns the algorithm name of this {@code KDF} object.
+     *
+     * @return the algorithm name of this {@code KDF} object
+     */
+    public String getAlgorithm() {
+        return this.algorithm;
+    }
+
+    /**
+     * Returns the name of the provider.
+     *
+     * @see <a href="#DelayedProviderSelection">Delayed Provider Selection</a>
+     *
+     * @return the name of the provider
+     */
+    public String getProviderName() {
+        useFirstSpi();
+        return pairOfSpiAndProv.provider().getName();
+    }
+
+    /**
+     * Returns the {@code KDFParameters} used with this {@code KDF} object.
+     * <p>
+     * The returned parameters may be the same that were used to initialize
+     * this {@code KDF} object, or may contain additional default or
+     * random parameter values used by the underlying KDF algorithm.
+     * If the required parameters were not supplied and can be generated by
+     * the {@code KDF} object, the generated parameters are returned;
+     * otherwise {@code null} is returned.
+     *
+     * @see <a href="#DelayedProviderSelection">Delayed Provider Selection</a>
+     *
+     * @return the parameters used with this {@code KDF} object, or
+     * {@code null}
+     */
+    public KDFParameters getParameters() {
+        useFirstSpi();
+        return pairOfSpiAndProv.spi().engineGetParameters();
+    }
+
+    /**
+     * Returns a {@code KDF} object that implements the specified algorithm.
+     *
+     * @see <a href="#DelayedProviderSelection">Delayed Provider Selection</a>
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use.
+     *     See the {@code KDF} section in the <a href=
+     *     "{@docRoot}/../specs/security/standard-names.html#kdf-algorithms">
+     *     Java Security Standard Algorithm Names Specification</a>
+     *     for information about standard KDF algorithm names.
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if no {@code Provider} supports a {@code KDF} implementation for the
+     *     specified algorithm
+     * @throws NullPointerException
+     *     if {@code algorithm} is {@code null}
+     */
+    public static KDF getInstance(String algorithm)
```

Comment:
One thing that is missing from this method is an Implementation Note describing how the `jdk.security.provider.preferred` property affects the search algorithm. See any other JCE API - you can just copy/paste the text. Also affects the `getInstance` method that takes an algorithm and params.